File: WordPress/Classes/ViewRelated/Themes/ThemeBrowserCell.m
#import "ThemeBrowserCell.h"
#import "Theme.h"
#import "WPImageSource.h"
#import "WPStyleGuide.h"

@interface WPImageSource (Theme)

- (void)downloadThemeScreenshotForURL:(NSURL*)url themeId:(NSString*)themeId success:(void(^)(UIImage *image, NSString *themeId))success failure:(void (^)(NSError *error))failure;

@end

@interface ThemeBrowserCell ()

@property (nonatomic, weak) UIImageView *screenshot;
@property (nonatomic, weak) UILabel *title;
@property (nonatomic, weak) UIImageView *statusIcon;

@end

@implementation ThemeBrowserCell

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        BOOL isRetina = [[UIApplication sharedApplication] respondsToSelector:@selector(scale)];
        self.contentView.layer.borderWidth = isRetina ? 0.5f : 1.0f;
        self.contentView.layer.borderColor = [[UIColor lightGrayColor] CGColor];
        self.contentView.backgroundColor = [UIColor colorWithWhite:0.95 alpha:1.0];

        UIImageView *screenshot = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.contentView.bounds.size.width, IS_IPAD ? 225.0f : 204.0f)];
        _screenshot = screenshot;
        [_screenshot setContentMode:UIViewContentModeScaleAspectFit];
        [self.contentView addSubview:_screenshot];

        // theme-browse-current/theme-browse-premium 37x37
        UIImageView *statusIcon = [[UIImageView alloc] initWithFrame:CGRectMake(self.contentView.frame.size.width - 37.0f, 0, 37.0f, 37.0f)];
        _statusIcon = statusIcon;
        [self.contentView addSubview:_statusIcon];

        UIView *titleContainer = [[UIView alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.screenshot.frame), self.contentView.bounds.size.width, 30.0f)];
        titleContainer.backgroundColor = [UIColor colorWithWhite:0.2 alpha:1.0];
        [self.contentView addSubview:titleContainer];

        UILabel *title = [[UILabel alloc] initWithFrame:CGRectMake(6, 0, titleContainer.frame.size.width-12, titleContainer.frame.size.height)];
        _title = title;
        _title.backgroundColor = [UIColor clearColor];
        _title.textColor = [UIColor whiteColor];
        _title.font = [WPStyleGuide regularTextFont];
        [titleContainer addSubview:_title];
    }
    return self;
}

- (void)prepareForReuse
{
    self.screenshot.image = nil;
    self.statusIcon.image = nil;
}

- (void)setTheme:(Theme *)theme
{
    _theme = theme;
    self.title.text = _theme.name;

    if (_theme.isPremium) {
        _statusIcon.image = [UIImage imageNamed:@"theme-browse-premium"];
    }
    if (_theme.isCurrentTheme) {
        _statusIcon.image = [UIImage imageNamed:@"theme-browse-current"];
    }

    [[WPImageSource sharedSource] downloadThemeScreenshotForURL:[NSURL URLWithString:self.theme.screenshotUrl] themeId:_theme.themeId success:^(UIImage *image, NSString *themeId) {
        if ([themeId isEqualToString:_theme.themeId]) {
            _screenshot.image = image;
        }
    } failure:^(NSError *error) {
        DDLogWarn(@"Theme screenshot failed to download for theme: %@ error: %@", _theme.themeId, error);
    }];
}

@end

@implementation WPImageSource (Theme)

// Save the themeId to avoid the scenario where the cell has already been reused and the theme is now different
- (void)downloadThemeScreenshotForURL:(NSURL *)url
                              themeId:(NSString*)themeId
                              success:(void (^)(UIImage *image, NSString *themeId))success
                              failure:(void (^)(NSError *))failure
{
    [self downloadImageForURL:url withSuccess:^(UIImage *image) {
        success(image, themeId);
    } failure:failure];
}

@end

File: WordPress/Classes/ViewRelated/Themes/ThemeBrowserCell.swift
import Foundation
import WordPressShared.WPStyleGuide
import CocoaLumberjack

/// Actions provided in cell button triggered action sheet
///
public enum ThemeAction {
    case activate
    case customize
    case details
    case support
    case tryCustomize
    case view

    static func activeActionsForTheme(_ theme: Theme) ->[ThemeAction] {
        if theme.custom {
            if theme.hasDetailsURL() {
                return [customize, details]
            }
            return [customize]
        }
        return [customize, details, support]
    }

    static func inactiveActionsForTheme(_ theme: Theme) ->[ThemeAction] {
        if theme.custom {
            if theme.hasDetailsURL() {
                return [tryCustomize, activate, details]
            }
            return [tryCustomize, activate]
        }
        return [tryCustomize, activate, view, details, support]
    }

    var title: String {
        switch self {
        case .activate:
            return NSLocalizedString("Activate", comment: "Theme Activate action title")
        case .customize:
            return NSLocalizedString("Customize", comment: "Theme Customize action title")
        case .details:
            return NSLocalizedString("Details", comment: "Theme Details action title")
        case .support:
            return NSLocalizedString("Support", comment: "Theme Support action title")
        case .tryCustomize:
            return NSLocalizedString("Try & Customize", comment: "Theme Try & Customize action title")
        case .view:
            return NSLocalizedString("View", comment: "Theme View action title")
        }
    }

    func present(_ theme: Theme, _ presenter: ThemePresenter) {
        switch self {
        case .activate:
            presenter.activateTheme(theme)
        case .customize:
            presenter.presentCustomizeForTheme(theme)
        case .details:
            presenter.presentDetailsForTheme(theme)
        case .support:
            presenter.presentSupportForTheme(theme)
        case .tryCustomize:
            presenter.presentPreviewForTheme(theme)
        case .view:
            presenter.presentViewForTheme(theme)
        }
    }
}

open class ThemeBrowserCell: UICollectionViewCell {
    // MARK: - Constants

    @objc public static let reuseIdentifier = "ThemeBrowserCell"

    // MARK: - Private Aliases

    fileprivate typealias Styles = WPStyleGuide.Themes

    // MARK: - Outlets

    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var infoBar: UIView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var infoLabel: UILabel!
    @IBOutlet weak var actionButton: UIButton!
    @IBOutlet weak var highlightView: UIView!
    @IBOutlet weak var activityView: UIActivityIndicatorView!

    // MARK: - Properties

    @objc open var theme: Theme? {
        didSet {
            refreshGUI()
        }
    }

    @objc open var showPriceInformation: Bool = false
    open weak var presenter: ThemePresenter?

    fileprivate var placeholderImage = UIImage(named: "theme-loading")
    fileprivate var activeEllipsisImage = UIImage(named: "icon-menu-ellipsis-white")
    fileprivate var inactiveEllipsisImage = UIImage(named: "icon-menu-ellipsis")

    // MARK: - GUI

    override open var isHighlighted: Bool {
        didSet {
            let alphaFinal: CGFloat = isHighlighted ? 0.3 : 0
            UIView.animate(withDuration: 0.2, animations: { [weak self] in
                self?.highlightView.alpha = alphaFinal
            })
        }
    }

    override open func awakeFromNib() {
        super.awakeFromNib()

        actionButton.isExclusiveTouch = true

        layer.borderWidth = Styles.cellBorderWidth
        infoBar.layer.borderWidth = Styles.cellBorderWidth
        nameLabel.font = Styles.cellNameFont
        infoLabel.font = Styles.cellInfoFont
        actionButton.layer.borderWidth = Styles.cellBorderWidth
    }

    override open func prepareForReuse() {
        super.prepareForReuse()
        theme = nil
        presenter = nil
        showPriceInformation = false
    }

    fileprivate func refreshGUI() {
        if let theme = theme {
            if let imageUrl = theme.screenshotUrl, !imageUrl.isEmpty {
                refreshScreenshotImage(imageUrl)
            } else {
                showPlaceholder()
            }

            nameLabel.text = theme.name
            if theme.isCurrentTheme() {
                backgroundColor = Styles.activeCellBackgroundColor
                infoBar.backgroundColor = Styles.activeCellBackgroundColor
                layer.borderColor = Styles.activeCellBorderColor.cgColor
                infoBar.layer.borderColor = Styles.activeCellDividerColor.cgColor
                actionButton.layer.borderColor = Styles.activeCellDividerColor.cgColor
                actionButton.setImage(activeEllipsisImage, for: .normal)

                nameLabel.textColor = Styles.activeCellNameColor
                infoLabel.textColor = Styles.activeCellInfoColor
                infoLabel.text = NSLocalizedString("ACTIVE", comment: "Label for active Theme browser cell")
            } else {
                backgroundColor = Styles.inactiveCellBackgroundColor
                infoBar.backgroundColor = Styles.inactiveCellBackgroundColor
                layer.borderColor = Styles.inactiveCellBorderColor.cgColor
                infoBar.layer.borderColor = Styles.inactiveCellDividerColor.cgColor
                actionButton.layer.borderColor = Styles.inactiveCellDividerColor.cgColor
                actionButton.setImage(inactiveEllipsisImage, for: .normal)

                nameLabel.textColor = Styles.inactiveCellNameColor
                if theme.isPremium() && showPriceInformation {
                    infoLabel.textColor = Styles.inactiveCellPriceColor
                    infoLabel.text = theme.price
                } else {
                    infoLabel.text = nil
                }
            }
        } else {
            imageView.image = nil
            nameLabel.text = nil
            infoLabel.text = nil
            activityView.stopAnimating()
        }

        prepareForVoiceOver()
    }

    fileprivate func showPlaceholder() {
        imageView.contentMode = .center
        imageView.backgroundColor = Styles.placeholderColor
        imageView.image = placeholderImage
        activityView.stopAnimating()
    }

    fileprivate func showScreenshot() {
        imageView.contentMode = .scaleAspectFit
        imageView.backgroundColor = UIColor.clear
        activityView.stopAnimating()
    }

    fileprivate func refreshScreenshotImage(_ imageUrl: String) {
        // Themes not hosted on WP.com have an incorrect screenshotUrl and do not correctly support the w param
        let imageUrlForWidth = imageUrl.hasPrefix("http") ? imageUrl + "?w=\(presenter!.screenshotWidth)" :
            String(format: "http:%@", imageUrl)
        let screenshotUrl = URL(string: imageUrlForWidth)

        imageView.backgroundColor = Styles.placeholderColor
        activityView.startAnimating()
        imageView.downloadImage(from: screenshotUrl, success: { [weak self] _ in
            self?.showScreenshot()
        }, failure: { [weak self] error in
                if let error = error as NSError?, error.domain == NSURLErrorDomain && error.code == NSURLErrorCancelled {
                    return
                }
                DDLogError("Error loading theme screenshot: \(String(describing: error?.localizedDescription))")
                self?.showPlaceholder()
        })
    }

    // MARK: - Actions

    @IBAction fileprivate func didTapActionButton(_ sender: UIButton) {
        guard let theme = theme, let presenter = presenter else {
            return
        }

        let alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)

        let themeActions = theme.isCurrentTheme() ? ThemeAction.activeActionsForTheme(theme) : ThemeAction.inactiveActionsForTheme(theme)
        themeActions.forEach { themeAction in
            alertController.addActionWithTitle(themeAction.title,
                                               style: .default,
                                               handler: { (action: UIAlertAction) in
                                                themeAction.present(theme, presenter)
            })
        }

        let cancelTitle = NSLocalizedString("Cancel", comment: "Cancel action title")
        alertController.addCancelActionWithTitle(cancelTitle, handler: nil)

        alertController.modalPresentationStyle = .popover
        alertController.presentFromRootViewController()

        if let popover = alertController.popoverPresentationController {
            popover.sourceView = actionButton
            popover.sourceRect = actionButton.bounds
            popover.permittedArrowDirections = .any
            popover.canOverlapSourceViewRect = true
        }
    }

}

extension ThemeBrowserCell: Accessible {
    func prepareForVoiceOver() {
        prepareCellForVoiceOver()
        prepareActionButtonForVoiceOver()
        prepareNameLabelForVoiceOver()
    }

    private func prepareCellForVoiceOver() {
        imageView.isAccessibilityElement = true
        if let name = theme?.name {
            imageView.accessibilityLabel = name
            imageView.accessibilityTraits = [.button, .summaryElement]
        }

        if let details = theme?.details {
            imageView.accessibilityHint = details
        }
    }

    private func prepareActionButtonForVoiceOver() {
        actionButton.isAccessibilityElement = true
        actionButton.accessibilityLabel = NSLocalizedString("More", comment: "Action button to display more available options")
        actionButton.accessibilityTraits = .button
    }

    private func prepareNameLabelForVoiceOver() {
        nameLabel.isAccessibilityElement = false
    }
}
